library/LuaTypes.cpp

```diff
@@ -1337,21 +1354,179 @@ static void IndexFields(lua_State *state, int base, struct_identity *pstruct, bo
+ *
+ * upvalue 1: name -> struct_field_info* table
+ * upvalue 3: field table (int <-> name)
+ */
+static int meta_fieldinfo_next(lua_State *state)
+{
+    if (lua_gettop(state) < 2) lua_pushnil(state);
+
+    int len = lua_rawlen(state, UPVAL_FIELDTABLE);
+    int idx = cur_iter_index(state, len+1, 2, 0);
+    if (idx == len)
+        return 0;
+
+    lua_rawgeti(state, UPVAL_FIELDTABLE, idx+1);
+
+    // modified from meta_struct_next:
+    // retrieve the struct_field_info* from the table and convert it
+    lua_dup(state);
+    lua_gettable(state, lua_upvalueindex(1));
+    auto field = static_cast<const struct_field_info*>(lua_touserdata(state, -1));
+    lua_pop(state, 1);
+    PushFieldInfoSubTable(state, field);
+
+    return 2;
+}
+
+static void AddFieldInfoTable(lua_State *state, int ftable_idx, struct_identity *pstruct)
+{
+    Lua::StackUnwinder base{state};
+
+    // metatable
+    lua_newtable(state);
+    int ix_meta = lua_gettop(state);
+
+    // field info table (name -> struct_field_info*)
+    lua_newtable(state);
+    int ix_fieldinfo = lua_gettop(state);
+
+    // field iter table (int <-> name)
+    lua_newtable(state);
+    int ix_fielditer = lua_gettop(state);
+    IndexFields(state, base, pstruct, IndexFieldsFlags::RAW);
+
+    PushStructMethod(state, ix_meta, ix_fielditer, meta_fieldinfo_next);
+    // change upvalue 1 to the field info table since we don't need the original
+    lua_pushvalue(state, ix_fieldinfo);
+    lua_setupvalue(state, -2, 1);
+    SetPairsMethod(state, ix_meta, "__pairs");
+
+    // field table (name -> table representation of struct_field_info)
+    lua_newtable(state);
+    int ix_fields = lua_gettop(state);
+
+    // wrapper table (empty, indexes into field table with metamethods)
+    lua_newtable(state);
+    int ix_wrapper = lua_gettop(state);
+
+    // set up metatable for the wrapper
+    // use field table for __index
+    lua_pushstring(state, "__index");
+    lua_pushvalue(state, ix_fieldinfo);
+    lua_pushcclosure(state, meta_fieldinfo_index, 1);
+    lua_settable(state, ix_meta);
+
+    // use change_error() for __newindex
+    lua_pushstring(state, "__newindex");
+    lua_getfield(state, LUA_REGISTRYINDEX, DFHACK_CHANGEERROR_NAME);
+    lua_settable(state, ix_meta);
+
+    lua_pushvalue(state, ix_meta);
+    lua_setmetatable(state, ix_wrapper);
+
+    // convert field info table (struct_field_info) to field table (lua tables)
+    lua_pushnil(state);  // initial key for next()
+    while (lua_next(state, ix_fieldinfo)) {
+        auto field = static_cast<const struct_field_info*>(lua_touserdata(state, -1));
+        lua_pushvalue(state, -2);  // field name
+        PushFieldInfoSubTable(state, field);
+        lua_settable(state, ix_fields);
+        lua_pop(state, 1); // struct_field_info
+    }
+
+    // lua_pushvalue(state, ix_fields);
+    // freeze_table(state); // TODO: figure out why this creates an __index cycle for nonexistent fields
```

Comment:
does the index cycle happen for other frozen tables, such as the `attrs` tables?

Reply:
Nope, but those define `__index` to be a metamethod, not a table.

I'm effectively freezing the table by setting `__newindex`, anyway, so I'm not really worried about this. It's a nice-to-have that could save maybe 5-10 lines of code, but now that I'm generating the field-specific tables on the fly, I only really need to freeze the outer table, so there's not much duplication.